trie/sync.py

```diff
@@ -0,0 +1,179 @@
+
+
+class HexaryTrieSync:
+
+    def __init__(self, root_hash, db, logger):
+        self.queue = []
+        self.requests = {}
+        self.db = db
+        self.root_hash = root_hash
+        self.logger = logger
+        self.schedule(root_hash, parent=None, depth=0, leaf_callback=self.leaf_callback)
+
+    def leaf_callback(self, data, parent):
+        """Called when we reach a leaf node.
+
+        Should be implemented by subclasses that need to perform special handling of leaves.
+        """
+
+    def next_batch(self, n=1):
+        """Return the next requests that should be dispatched."""
+        if len(self.queue) == 0:
+            return []
+        batch = self.queue[-n:]
+        self.queue = self.queue[:-n]
+        return list(reversed(batch))
+
+    def schedule(self, node_key, parent, depth, leaf_callback, is_raw=False):
+        """Schedule a request for the node with the given key."""
+        request = SyncRequest(node_key, parent, depth, leaf_callback, is_raw)
+        if request.node_key in self.db:
+            self.logger.debug("Node %s already exists in db" % encode_hex(request.node_key))
+            return
+
+        for ancestor in request.parents:
+            ancestor.dependencies += 1
+
+        existing = self.requests.get(request.node_key)
+        if existing is not None:
+            self.logger.debug(
+                "Already requesting %s, will just update parents list" % request)
+            existing.parents.extend(request.parents)
+            return
+
+        # Requests get added to both self.queue and self.requests; the former is used to keep
+        # track which requests should be sent next, and the latter is used to avoid scheduling a
+        # request for a given node multiple times.
+        self.requests[request.node_key] = request
+        bisect.insort(self.queue, request)
+
+    def get_children(self, request):
+        """Return all children of the node retrieved by the given request.
+
+        :rtype: A two-tuple with one list containing the children that reference other nodes and
+        another containing the leaf children.
+        """
+        node = decode_node(request.data)
+        node_type = get_node_type(node)
+        references = []
+        leaves = []
+        if node_type == NODE_TYPE_LEAF:
+            leaves.append(node[1])
+        elif node_type == NODE_TYPE_EXTENSION:
+            depth = request.depth + len(node[0])
+            references.append((depth, node[1]))
+        elif node_type == NODE_TYPE_BRANCH:
+            depth = request.depth + 1
+            for item in node[:16]:
+                if is_blank_node(item):
+                    continue
+
+                # In a branch, the first 16 items are either a node whose RLP-encoded
+                # representation is under 32 bytes or a reference to another node.
+                if len(item) == 2:
+                    assert get_node_type(item) == NODE_TYPE_LEAF
```

Comment:
*nitpick* 

I'd be more comfortable with explicit exceptions over assertions in productions code since assertions are easy to disable.